src/subscription.js

```diff
@@ -1040,16 +1046,11 @@ Subscription.prototype.openConnection_ = function() {
   });
 
   pool.on('message', function(message) {
-    if (!self.hasMaxMessages_()) {
-      self.emit('message', self.leaseMessage_(message));
-      return;
-    }
+    self.emit('message', self.leaseMessage_(message));
 
-    if (!pool.isPaused) {
+    if (!pool.isPaused && self.hasMaxMessages_()) {
       pool.pause();
     }
-
-    message.nack();
```

Comment:
Per the PubSub team nacking messages like this can lead to an increased number of duplicate messages. Since this PRs main focus is on refactoring flow control, we can remove this line and will hopefully see fewer duplicates.